instrumentation/apache-dubbo-2.7/testing/src/main/java/io/opentelemetry/instrumentation/apachedubbo/v2_7/AbstractDubboTest.java

```diff
@@ -0,0 +1,290 @@
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInstance;
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+abstract class AbstractDubboTest {
+
+  private final ProtocolConfig protocolConfig = new ProtocolConfig();
+
+  abstract InstrumentationExtension testing();
+
+  private DubboBootstrap bootstrap;
+  private DubboBootstrap consumerBootstrap;
+
+  @BeforeAll
+  static void setUp() throws Exception {
+    System.setProperty("dubbo.application.qos-enable", "false");
+    Field field = NetUtils.class.getDeclaredField("LOCAL_ADDRESS");
+    field.setAccessible(true);
+    field.set(null, InetAddress.getLoopbackAddress());
+  }
+
+  @AfterAll
+  static void tearDown() {
+    System.clearProperty("dubbo.application.qos-enable");
+  }
+
+  @AfterEach
+  @SuppressWarnings("CatchingUnchecked")
+  public void afterEach() {
+    List<DubboBootstrap> dubboBootstraps = Arrays.asList(bootstrap, consumerBootstrap);
+    for (DubboBootstrap bootstrap : dubboBootstraps) {
+      try {
+        if (bootstrap != null) {
+          bootstrap.destroy();
+        }
+      } catch (Exception e) {
+        // ignore
+      }
+    }
+  }
+
+  ReferenceConfig<HelloService> configureClient(int port) {
+    ReferenceConfig<HelloService> reference = new ReferenceConfig<>();
+    reference.setInterface(HelloService.class);
+    reference.setGeneric("true");
+    reference.setUrl("dubbo://localhost:" + port + "/?timeout=30000");
+    return reference;
+  }
+
+  ServiceConfig<HelloServiceImpl> configureServer() {
+    RegistryConfig registerConfig = new RegistryConfig();
+    registerConfig.setAddress("N/A");
+    ServiceConfig<HelloServiceImpl> service = new ServiceConfig<>();
+    service.setInterface(HelloService.class);
+    service.setRef(new HelloServiceImpl());
+    service.setRegistry(registerConfig);
+    return service;
+  }
+
+  @Test
+  @SuppressWarnings({"rawtypes", "unchecked"})
+  void testApacheDubboBase() {
+    int port = PortUtils.findOpenPort();
+    protocolConfig.setPort(port);
+    // provider boostrap
+    bootstrap = DubboTestUtil.newDubboBootstrap();
+    bootstrap
+        .application(new ApplicationConfig("dubbo-test-provider"))
+        .service(configureServer())
+        .protocol(protocolConfig)
+        .start();
+
+    // consumer boostrap
+    consumerBootstrap = DubboTestUtil.newDubboBootstrap();
+    ReferenceConfig referenceConfig = configureClient(port);
+    ProtocolConfig consumerProtocolConfig = new ProtocolConfig();
+    consumerProtocolConfig.setRegister(false);
+    consumerBootstrap
+        .application(new ApplicationConfig("dubbo-demo-api-consumer"))
+        .reference(referenceConfig)
+        .protocol(consumerProtocolConfig)
+        .start();
+
+    // generic call
+    ReferenceConfig<GenericService> reference = referenceConfig;
+    GenericService genericService = reference.get();
+
+    Object[] o = new Object[1];
+    o[0] = "hello";
```

Comment:
I'd just inline `new Object[] { "hello" }` in `genericService.$invoke` parameters